### PR TITLE
8264696: Multi-catch clause causes compiler exception because it uses the package-private supertype

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

The method `isAccessible(Env<AttrContext> env, Type t, boolean checkInner)` should take `UnionClassType` as a special type to do some corresponding work. If not, the problem will occur. This patch fixes it and adds a test case.

Thank you for taking the time to review.

Best Regards.
-- Guoxiong

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264696](https://bugs.openjdk.java.net/browse/JDK-8264696): Multi-catch clause causes compiler exception because it uses the package-private supertype

### Reviewers
 * [Vicente Romero](https://openjdk.java.net/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3374/head:pull/3374` \
`$ git checkout pull/3374`

Update a local copy of the PR: \
`$ git checkout pull/3374` \
`$ git pull https://git.openjdk.java.net/jdk pull/3374/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3374`

View PR using the GUI difftool: \
`$ git pr show -t 3374`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3374.diff">https://git.openjdk.java.net/jdk/pull/3374.diff</a>

</details>

### CODE REVIEW

File: test/langtools/tools/javac/resolve/PackagePrivateSupertypeAtMultiCatch.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    public void testPackagePrivateSupertypeAtMultiCatch() throws Exception {
+        Path src = Paths.get("src");
+        tb.writeJavaFiles(src, testCode, parentExceptionCode, child1ExceptionCode, child2ExceptionCode);
+        Path out = Paths.get("out");
+        Files.createDirectories(out);
+        new JavacTask(tb)
+                .files(tb.findJavaFiles(src))
+                .outdir(out)
+                .run()
+                .writeAll();
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/resolve/PackagePrivateSupertypeAtMultiCatch.java:105]: nit: you don't need the `writeAll()` method given that you are not collecting the output

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java [hunk 2]
```diff
@@ -385,9 +385,17 @@ boolean isAccessible(Env<AttrContext> env, Type t) {
     }
 
     boolean isAccessible(Env<AttrContext> env, Type t, boolean checkInner) {
-        return (t.hasTag(ARRAY))
-            ? isAccessible(env, types.cvarUpperBound(types.elemtype(t)))
-            : isAccessible(env, t.tsym, checkInner);
+        if (t.hasTag(ARRAY)) {
+            return isAccessible(env, types.cvarUpperBound(types.elemtype(t)));
+        } else if (!t.isUnion()) {
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:390]: I would change the code a bit, like in:

```
        } else if (t.isUnion()) {
            return StreamSupport.stream(((UnionClassType) t).getAlternativeTypes().spliterator(), false)
                    .allMatch(alternative -> isAccessible(env, alternative.tsym, checkInner));
        } else {
            return isAccessible(env, t.tsym, checkInner);
        }
```

**Review by vicente-romero-oracle — CHANGES_REQUESTED** [review]

File: test/langtools/tools/javac/resolve/PackagePrivateSupertypeAtMultiCatch.java (hunk 1 above)
Comment (lgxbslgx) [reply test/langtools/tools/javac/resolve/PackagePrivateSupertypeAtMultiCatch.java:105]: Fixed.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java (hunk 2 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java]: Very nice suggestion. Fixed.

**Review by vicente-romero-oracle — APPROVED** [review]
looks good, thanks for the fix

**Pull request closed without merging**